### PR TITLE
Word lo/hi refactor

### PR DESCRIPTION
**Pull request by ed255**
Update all specs to replace word_rlc by word hi/lo, following the proposal presented here https://github.com/privacy-scaling-explorations/zkevm-specs/issues/363

- [x] bytecode
- [x] copy_circuit
- [x] exp_circuit
- [x] public_inputs
- [x] state
- [x] tx
- [x] evm

Some notes about the code refactor:
- Introduced a new type `Word` that holds `lo, hi` as `Expression`s.  This type has some methods to work with Words without needing to use the internal `lo, hi` fields, like the `==` operator to compare words, or the `|` operator to "merge" words.
- Introduced new methods in `Instruction` to be used with Word, most of them replace `rlc` corresponding methods.
- Introduced a new type `WordOrValue` that wraps a `Word` to be used in tables where depending on the tag the value is a word or a single value.  This type remembers if the internal content fits in a single field value: this is not a constraint, but helps catching bugs in the spec (when accessing the single field value via `WordOrValue.value()`.
- Previously table rows were defined as a list of `Expression`.  I updated it to have a mix of `Expression` and `Word` to have more readable code (instead of explicitly having word_lo, word_hi in the rows).  This means that all the logic for lookups now can have a mix of Expression and Word.
- Removed all the randomness values used in tests except for the randomness_keccak used in the Keccak Circuit
- In general, everything that previously was `RLC(foo, n_bytes=32)` is now `Word(foo)`
- Added file `tests/conftest.py` with a setting that will make pytest rewrite the asserts in `src/zkevm_specs/` so that they print the values used in the assertions automatically.

### CODE REVIEW

File: src/zkevm_specs/state_circuit.py [hunk 1]
```diff
@@ -171,49 +206,54 @@ def assert_in_range(x: FQ, min_val: int, max_val: int) -> None:
 
 @is_circuit_code
 def check_start(row: Row, row_prev: Row):
+    print(row)
```
Comment (han0110) [inline src/zkevm_specs/state_circuit.py:209]:
```suggestion
```

File: src/zkevm_specs/util/constraint_system.py [hunk 2]
```diff
@@ -34,11 +34,22 @@ def constrain_equal(self, lhs: Expression, rhs: Expression):
...
+
     def constrain_zero(self, value: Expression):
         assert self._eval(value) == 0, ConstraintUnsatFailure(
             f"Expected value to be 0, but got {value}"
         )
 
+    def constrain_zero_word(self, value: Word):
+        assert self._eval(value.lo.expr()) == 0 and self._eval(
+            value.hi.expr()
+        ), ConstraintUnsatFailure(f"Expected word to be 0, but got {value}")
```
Comment (han0110) [inline src/zkevm_specs/util/constraint_system.py:51]:
```suggestion
        assert (
            self._eval(value.lo.expr()) == 0 and self._eval(value.hi.expr()) == 0
        ), ConstraintUnsatFailure(f"Expected word to be 0, but got {value}")
```

**Review by han0110 — COMMENTED** [review]
First pass

File: specs/tables.md [hunk 3]
```diff
@@ -4,27 +4,33 @@ For the zkevm we use the following dynamic and fixed tables for lookups to the E
...
+| $TxID    | Gas                 | 0          | $value,0       |
+| $TxID    | GasPrice            | 0          | $value{Lo,Hi}  |
+| $TxID    | CallerAddress       | 0          | $value,0       |
+| $TxID    | CalleeAddress       | 0          | $value,0       |
+| $TxID    | IsCreate            | 0          | $value,0       |
+| $TxID    | Value               | 0          | $value{Lo,Hi}  |
+| $TxID    | CallDataLength      | 0          | $value,0       |
+| $TxID    | CallDataGasCost     | 0          | $value,0       |
+| $TxID    | TxSignHashLo        | 0          | $value,0       |
+| $TxID    | TxSignHashHi        | 0          | $value,0       |
```
[the remark this answers is not in the document]
Comment (ed255) [reply specs/tables.md]: Fixed in https://github.com/privacy-scaling-explorations/zkevm-specs/pull/390/commits/82919ed109795b9c575182f2703ba268dfd0f91e

File: specs/tables.md [hunk 4]
```diff
@@ -73,82 +79,84 @@ The correctness of the rw_table is validated in the state circuit.
...
+| $counter | true        | TxAccessListAccountStorage | $txID     | $address           |                            | $storageKey{Lo,Hi}        | $val,0           | $valPrev,0            |                        |
+| $counter | $isWrite    | TxRefund                   | $txID     |                    |                            |                           | $val,0           | $valPrev,0            |                        |
+|          |             |                            |           |                    |                            |                           |                  |                       |                        |
+|          |             |                            |           |                    | *AccountFieldTag*          |                           |                  |                       |                        |
+| $counter | $isWrite    | Account                    |           | $address           | Nonce                      |                           | $val,0           | $valPrev,0            | $committedValue{Lo,Hi} |
```
[the remark this answers is not in the document]
Comment (ed255) [reply specs/tables.md]: Fixed in https://github.com/privacy-scaling-explorations/zkevm-specs/pull/390/commits/82919ed109795b9c575182f2703ba268dfd0f91e

File: specs/tables.md [hunk 5]
```diff
@@ -204,39 +212,31 @@ __Hence the addition inside of the block_table.__
...
+| -------   | ----------------------- | ----         | -------------        | ------------        | ---------         | -----         |
+| $addr     | NonceMod                | 0,0          | $noncePrev,0         | $nonceCur,0         | $rootPrev{Lo,Hi}  | $root{Lo,Hi}  |
+| $addr     | BalanceMod              | 0,0          | $balancePrev{Lo,Hi}  | $balanceCur{Lo,Hi}  | $rootPrev{Lo,Hi}  | $root{Lo,Hi}  |
+| $addr     | CodeHashMod             | 0,0          | $codeHashPrev{Lo,Hi} | $codeHashCur{Lo,Hi} | $rootPrev{Lo,Hi}  | $root{Lo,Hi}  |
+| $addr     | NonExistingAccountProof | 0,0          | 0,0                  | 0,0                 | $rootPrev{Lo,Hi}  | $root{Lo,Hi}  |
+| $addr     | AccountDeleteMod        | 0,0          | 0,0                  | 0,0                 | $rootPrev{Lo,Hi}  | $root{Lo,Hi}  |
+| $addr     | StorageMod              | $key         | $valuePrev{Lo,Hi}    | $value{Lo,Hi}       | $rootPrev{Lo,Hi}  | $root{Lo,Hi}  |
+| $addr     | NonExistingStorageProof | $key         | 0,0                  | 0,0                 | $rootPrev{Lo,Hi}  | $root{Lo,Hi}  |
```
[the remark this answers is not in the document]
Comment (ed255) [reply specs/tables.md]: Fixed in https://github.com/privacy-scaling-explorations/zkevm-specs/pull/390/commits/82919ed109795b9c575182f2703ba268dfd0f91e

File: src/zkevm_specs/state_circuit.py (hunk 1 above)
Comment (ed255) [reply src/zkevm_specs/state_circuit.py]: Fixed in https://github.com/privacy-scaling-explorations/zkevm-specs/pull/390/commits/82919ed109795b9c575182f2703ba268dfd0f91e

**Review by han0110 — APPROVED** [review]
Final nits, other look good to me! Nice work!

File: src/zkevm_specs/evm_circuit/table.py [hunk 6]
```diff
@@ -507,10 +518,12 @@ class ExpTableRow(TableRow):
...
-    exponent_lo: FQ
-    exponent_hi: FQ
-    exponentiation_lo: FQ
-    exponentiation_hi: FQ
+    exponent: Word
+    # exponent_lo: FQ
+    # exponent_hi: FQ
+    exponentiation: Word
+    # exponentiation_lo: FQ
+    # exponentiation_hi: FQ
```
[the remark this answers is not in the document]
Comment (ed255) [reply src/zkevm_specs/evm_circuit/table.py]: Fixed in https://github.com/privacy-scaling-explorations/zkevm-specs/pull/390/commits/01489b686314a0611e240549ece453bf8d92c1e5

File: src/zkevm_specs/util/arithmetic.py [hunk 7]
```diff
@@ -73,10 +77,108 @@ def __repr__(self) -> str:
...
+    def __repr__(self) -> str:
+        return f"Word({hex(self.int_value())})"
+
+    def __eq__(self, other) -> bool:
+        assert isinstance(other, Word)
+        return self.lo.expr() == other.lo.expr() and self.hi.expr() == other.hi.expr()
+
+    def __or__(self, other: Word) -> Word:
+        """Combine two words by adding their corresponding lo and hi parts.  Useful with select"""
+        return Word((self.lo.expr() + other.lo.expr(), self.hi.expr() + other.hi.expr()))
```
[the remark this answers is not in the document]
Comment (ed255) [reply src/zkevm_specs/util/arithmetic.py]: You're right, `__add__` seems more appropriate for this!  Fixed in https://github.com/privacy-scaling-explorations/zkevm-specs/pull/390/commits/01489b686314a0611e240549ece453bf8d92c1e5

**Review by hero78119 — COMMENTED** [review]
First review pass and it's really awesome job !!! 💯  

Just have few comments

1. Was thinking to keep `WordOrValue` definition in table, but do not expose it to any lookup method return type. Instead, we define separate lookup method to return either `Expression` or `Word` explicitly. Instead, we can have `instruction.xxxx_lookup() -> Expression` and `instruction.xxxx_lookup_word() -> Word` to return specific type directly. It also more close to the circuit part usage.

2. One readibility improvement, for `Expression.expr()` return type is FQ. Will it be better to rename `expr()` to `fq()`? 

3. maybe can define `Word.lo(FQ(xxx))` classmathod so can omit hi FQ(0) part. For example
```python
instruction.constrain_equal_word(
        Word((instruction.call_context_lookup(CallContextFieldTag.CallDataLength).value(), FQ(0)))
```
to 
```python
Word.lo(
    instruction.call_context_lookup(CallContextFieldTag.CallDataLength).value()
)
```

Other nitpicks
- Few FQ(0) can be FQ.zero() and Word(0) to Word.zero() instead 😃

**Comment by ed255** [comment]
> First review pass and it's really awesome job !!! 100
> 
> Just have few comments

Thanks for all the suggestions!

>     1. Was thinking to keep `WordOrValue` definition in table, but do not expose it to any lookup method return type. Instead, we define separate lookup method to return either `Expression` or `Word` explicitly, like `instruction.xxxx_lookup() -> Expression` and `instruction.xxxx_lookup_word() -> Word` to return specific type directly. It also more close to the circuit part usage.

I've added the functions `instruction.xxx_lookup_word()` so that the opcode implementations don't need to deal with WordOrValue directly.  Nevertheless the return type of `instruction.xxx_lookup_word()` is still `WordOrValue` because it allows some code deduplication, and `WordOrValue` inherits from `Word` so it can just be used as `Word` in the opcode implementations.

>     2. One readibility improvement, for `Expression.expr()` return type is FQ. Will it be better to rename `expr()` to `fq()`?

I think the reason to have `Expression.expr()` is to mimick the Rust code.  In the python implementation we're always evaluating the expressions so we're actually working with `FQ` all the time, but the code mimicks the Rust code where we are working with expressions.  So I think having `.expr()` makes it easier to map the python spec to the Rust implementation.  Also this would be a significant change, so if you think we shoud do this I suggest to address it in another PR :)

>     3. maybe can define `Word.lo(FQ(xxx))` classmathod so can omit hi FQ(0) part. For example
> 
> 
> ```python
> Word(
>      (instruction.call_context_lookup(CallContextFieldTag.CallDataLength).value(), FQ(0)
> )
> ```
> 
> to
> 
> ```python
> Word.lo(
>     instruction.call_context_lookup(CallContextFieldTag.CallDataLength).value()
> )
> ```

I've added a classmethod `from_lo`.  I didn't pick `lo` to avoid confusion with the `Word` class field `lo`.

> Other nitpicks
> 
>     * Few FQ(0) can be FQ.zero() and Word(0) to Word.zero() instead smiley

Ah, I think `FQ(0)` and `Word(0)` is more readable than `FQ.zero()` or `Word.zero()`.  Also we have a lot of `FQ(0)` in the code, so if you think this is a change that we should consider I would ask you to open an issue to avoid increasing the size of this PR.

**Pull request merged**